### PR TITLE
Added rule rest endpoint and backend operation for rule execution simulation

### LINKED ISSUE #2044: Ephemeris conditions not evaluated for schedule
Hey,

when configuring rules with ephemeris conditions, this are not evaluated when displaying the rules in the schedule - which makes the schedule incorrect and useless because the information displayed are wrong :(

When configuring the rule with an "DayOfWeekCondition" like so:
```
triggers:
  - id: "1"
    configuration:
      time: 10:00
    type: timer.TimeOfDayTrigger
conditions:
  - inputs: {}
    id: "2"
    configuration:
      days:
        - SAT
        - SUN
    type: timer.DayOfWeekCondition
actions: []

```

It will be displayed correct within the schedule:
Schedule_daysOfWeek

When using the "WeekendCondition"
```
triggers:
  - id: "1"
    configuration:
      time: 10:00
    type: timer.TimeOfDayTrigger
conditions:
  - inputs: {}
    id: "2"
    configuration: {}
    type: ephemeris.WeekendCondition
actions: []
```
the rule will be displayed all days:
Schedule_weekend

The epemeris configuration is set as saturday and sunday to be weekend:
ephemeris_settings

This applies to all the ephemeris conditions - like holidays too.

### LINKED ISSUE #2266: [automation] Schedule shows disabled rules
The schedule displays all rules - independent of the state. Event when rules are disabled by user they will be displayed in the schedule - this is really confusing.

### PR DESCRIPTION
**Pull request by soenkekueper**
Fixes #2044
Fixes #2266

UI PR is openhab/openhab-webui#815

### CODE REVIEW

**Comment by soenkekueper** [comment]
Hey,

i need your help here :(

The jenkins build fails with the following error messages:
`[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:58 min (Wall Clock)
[INFO] Finished at: 2021-01-19T01:41:36Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project org.openhab.core.automation.integration.tests: Compilation failure: Compilation failure: 
[ERROR] /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/itests/org.openhab.core.automation.integration.tests/src/main/java/org/openhab/core/automation/integration/test/RuleSimulationTest.java:[41,8] The import org.openhab.core.automation.RuleExecution cannot be resolved
[ERROR] /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/itests/org.openhab.core.automation.integration.tests/src/main/java/org/openhab/core/automation/integration/test/RuleSimulationTest.java:[135,14] RuleExecution cannot be resolved to a type
[ERROR] /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/itests/org.openhab.core.automation.integration.tests/src/main/java/org/openhab/core/automation/integration/test/RuleSimulationTest.java:[135,53] The method simulateRuleExecutions(java.time.ZonedDateTime, java.time.ZonedDateTime) is undefined for the type org.openhab.core.automation.RuleManager
[ERROR] /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/itests/org.openhab.core.automation.integration.tests/src/main/java/org/openhab/core/automation/integration/test/RuleSimulationTest.java:[141,18] RuleExecution cannot be resolved to a type
[ERROR] /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/itests/org.openhab.core.automation.integration.tests/src/main/java/org/openhab/core/automation/integration/test/RuleSimulationTest.java:[162,40] RuleExecution cannot be resolved to a type
[ERROR] /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/itests/org.openhab.core.automation.integration.tests/src/main/java/org/openhab/core/automation/integration/test/RuleSimulationTest.java:[191,47] CFG_START_TIME cannot be resolved or is not a field
[ERROR] /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/itests/org.openhab.core.automation.integration.tests/src/main/java/org/openhab/core/automation/integration/test/RuleSimulationTest.java:[192,47] CFG_END_TIME cannot be resolved or is not a field
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :org.openhab.core.automation.integration.tests
Build step 'Execute shell' marked build as failure
[htmlpublisher] Archiving HTML reports...
[htmlpublisher] Archiving at BUILD level /home/jenkins/jenkins-agent1/workspace/PR-openHAB-Core/target to /var/jenkins_home/jobs/PR-openHAB-Core/builds/3049/htmlreports/Static_20Code_20Analysis_20Report
Finished: FAILURE`

So this seems that my changes within the bundle _org.openhab.core.automation_ are not available when building the _org.openhab.core.automation.integration.tests_ bundle. It seems to me if an wrong version of the previous build bundle has been resolved?!

Confusing is also that the jenkins status page https://ci.openhab.org/job/PR-openHAB-Core/3049/ shows up green for all the given modules but the build is failed...

What can i do?

Of course this compiles and runs on my machine :D

Thanks!

**Comment by ghys** [comment]
@openhab/core-maintainers maybe you'd want to have a look at this PR?
Overall it looks good, and despite being API breaking, along with https://github.com/openhab/openhab-webui/pull/815, I think it would be a good addition to have a more accurate "Schedule" page in the UI (it would take Ephemeris into account etc.)

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/RuleManager.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/RuleManager.java:103]: Can you add JavaDoc for both parameters and rename this one to `until`:

```suggestion
     * @param until {@link ZonedDateTime} until the executions should be simulated.
```

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/RuleExecutionSimulator.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/RuleExecutionSimulator.java:66]: Add param `fromDate`.

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/GenericCronTriggerHandler.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/GenericCronTriggerHandler.java:88]: Is this annotation correct? I mean the new interface uses `@Nullable`.

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/handler/TimeBasedTriggerHandler.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/handler/TimeBasedTriggerHandler.java:31]: Minor reordering of method modifiers:

```suggestion
    abstract @Nullable SchedulerTemporalAdjuster getTemporalAdjuster();
```

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/TimeOfDayTriggerHandler.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/TimeOfDayTriggerHandler.java:96]: Correct annotation?

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/CronAdjuster.java
Comment (cweitkamp) [inline bundles/org.openhab.core/src/main/java/org/openhab/core/common/CronAdjuster.java:13]: Imo `org.openhab.core.scheduler` is the better place to move this class. Isn it?

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/RuleExecutionSimulator.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/RuleExecutionSimulator.java:87]: I am wondering if this method should be part of the public API too. It seems to make sense to just simulate a single rule too. We maybe can add a dedicated REST endpoint too.

File: bundles/org.openhab.core.automation.rest/src/main/java/org/openhab/core/automation/rest/internal/RuleResource.java
Comment (cweitkamp) [inline bundles/org.openhab.core.automation.rest/src/main/java/org/openhab/core/automation/rest/internal/RuleResource.java:349]: Only a guess - I leave it up to you to change it. The method returns a list of rule simulations filtered by tag "Schedule". The description is not 100% clear and users might be confused. So we either could clarify it by defining an own path for "Scheduler" or add some more information in the description.

e.g.

```suggestion
    @Path("/scheduler/simulations")
    @Produces(MediaType.APPLICATION_JSON)
    @Operation(operationId = "getSchedulerRuleSimulations", summary = "Simulates the executions of rules filtered by tag 'Schedule'.", responses = {
```

Would it make sense to add the "tags" parameter?

**Review by cweitkamp — COMMENTED** [review]
Thanks for your work. I looked into your PR and left some thoughts.

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/GenericCronTriggerHandler.java
Comment (soenkekueper) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/GenericCronTriggerHandler.java:88]: Yes you're right, in TimeBasedTriggerHandler i've also used `@Nullable` but this isn't needed at all, so i've removed it at all places.

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/RuleManager.java
Comment (soenkekueper) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/RuleManager.java:103]: I've renamed all from untilDate to until and added missing documentation.

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/TimeOfDayTriggerHandler.java
Comment (soenkekueper) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/module/handler/TimeOfDayTriggerHandler.java:96]: Removed annotation, due not needed.

File: bundles/org.openhab.core/src/main/java/org/openhab/core/common/CronAdjuster.java
Comment (soenkekueper) [reply bundles/org.openhab.core/src/main/java/org/openhab/core/common/CronAdjuster.java:13]: Good idea!

File: bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/RuleExecutionSimulator.java
Comment (soenkekueper) [reply bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/RuleExecutionSimulator.java:87]: I prefer the YAGNI principle, its not needed at the moment, so if this changes it can easy be made accessible.

File: bundles/org.openhab.core.automation.rest/src/main/java/org/openhab/core/automation/rest/internal/RuleResource.java
Comment (soenkekueper) [reply bundles/org.openhab.core.automation.rest/src/main/java/org/openhab/core/automation/rest/internal/RuleResource.java:349]: I've changed the rest endpoint according to your suggestion.
To allow filtering in simulation may be an good idea but would require additional work in web ui, so i would add the parameter when its needed.

**Review by cweitkamp — APPROVED** [review]
LGTM. Thank you.

**Pull request merged**
